(PUP-5756) remove dependency on eventlog

MosesMendoza opened the pull request:
This re-opens https://github.com/puppetlabs/puppet/pull/5509, which was reverted due to CI failures when ruby returned a binary string that originated in Encoding::Windows_31J that we attempted to report (see the last comments in PUP-5756).

This PR adds two new commits on top of the previous PR, one that improves our handling of conversion of BINARY strings in Puppet::Util::CharacterEncoding, and a second to leverage this in our EventLog reporting. A different approach would be to update the base Log class, but given CharacterEncoding hasn't been out in the wild yet I'm a little wary of patching everything to use it - one logging class felt like a reasonable test given this is where we saw the original error.

    (PUP-5756) (PUP-7021) better handling of BINARY string conversion

    Prior to this commit, we would transcode to UTF-8 a string not composed of
    valid UTF-8 bytes but valid in its current encoding. This presents a challenge
    for BINARY strings, which if originating in an encoding other to UTF-8 are
    otherwise valid, but cannot be transcoded (because we have no idea where they
    came from). In order to transcode successfully, we need an originating
    encoding to work with. Rather than just blindly call encode! with binary
    strings, we can take one more guess as to their origin - which is that they
    originated in `Encoding.default_external`. I.e., if a system is running in
    Encoding::Windows_31J, and a string is returned as BINARY, we can guess that
    the string might have been Encoding::Windows_31J. This perhaps substantially
    increases our odds that we'll be able to conver the (binary) string to UTF-8
    successfully.

and

    (PUP-5756) convert eventlog entries to UTF-8

    This commit updates the EventLog destination to utilize
    Puppet::Util::CharacterEncoding to convert messages to UTF-8. This sanitization
    will be particularly helpful in the case where we get back messages from ruby
    in BINARY that we can't convert cleanly to UTF-16LE without some intermediate
    conversions.

puppetcla commented:
CLA signed by all contributors.

Iristyle commented:
@MosesMendoza looks like we have some test failures on Travis

MosesMendoza commented:
Ha interesting. I'll dig into it. It appears at least from the way travis is rendering the output that it's getting what it expects but still failing on each test.
```
       expected: "\xFE\xF1"
            got: "\xFE\xF1"
```

Iristyle commented:
@MosesMendoza yeah, I noticed that... I wonder if the metadata on the string has a different encoding?

MosesMendoza commented:
Yeah, exactly. I found it - side-effect of the latest change affects the original string's external encoding rather than leaving it alone on failure

MosesMendoza commented:
Updated. Added a new commit c4e32cf that address previous side-effect. Will squash prior to merge, separate for review.

Also kicked off an ad-hoc build with all arch/version/locale windows test targets represented, will post results on completion

MosesMendoza commented on spec/unit/etc_spec.rb:
```diff
@@ -50,39 +51,85 @@
+
+      after(:each) do
+        Encoding.default_external = original_encoding
+      end
```
We end up repeating this pattern a few times, and I wonder if it's going to come up more.

Might be worth putting this into a spec util helper function, something callable like
```
with_external_encoding(Encoding::EUC_KR) do
  let(:group) { Etc::Group.new }
  ...
```

Iristyle replied:
I like that idea

joshcooper commented on lib/puppet/util/character_encoding.rb:
```diff
@@ -62,13 +63,27 @@ def convert_to_utf_8!(string)
       rescue EncodingError => detail
+        # Ensure the string retains it original external encoding since
+        # we've failed
+        string.force_encoding(original_encoding)
```
This method modifies the `string` parameter and also returns a value. Can we make it do one or the other?

MosesMendoza replied:
@joshcooper I'm sorry I'm not connecting the dots. Are you saying this line would end up modifying the string from its original state? My hope was that it was actually doing the opposite - restoring its original state after it was modified on line 76 in anticipation of calling `encode!`. I could modify line 86 to only do `force_encoding` if the same criteria are satisfied as we demand on line 76, ie only if the string's original encoding was BINARY and default external is not UTF-8? Is that what you mean?

joshcooper replied:
More a general comment about the `convert_to_utf8!` method. We generally modify the string object that is passed in, and then we return that same object, e.g. [`return string.encode!(...)](https://github.com/MosesMendoza/puppet/blob/c4e32cfdb25a87da21048f19d65402a272b94b4a/lib/puppet/util/character_encoding.rb#L77). But we also `return nil`](https://github.com/MosesMendoza/puppet/blob/c4e32cfdb25a87da21048f19d65402a272b94b4a/lib/puppet/util/character_encoding.rb#L92). So the API sometimes modifies the string that was passed in, and sometimes returns a nil value. As a consumer of the API, I'm not sure which value I should use after the API is called (the string that was passed in or the return value). Can the method be simplified to either always return the "resulting" object (making a copy of the original), or always modify the passed in object and not returning a value?

MosesMendoza replied:
Ah that makes sense, thanks. Yes I agree with your concern. In its first form the util never modified the original string and always returned a copy. The performance implications of always copying the string (in some cases multiple times) were not insignificant, so we adjusted to modifying the string in place. But, as you point out, we ended up with inconsistent return semantics. I'm inclined to modify the string in place and always return nil to keep the performance improvement but improve consistency.

MosesMendoza commented:
re-added blocked label as this is undergoing further revision based on feedback

MosesMendoza commented:
@Iristyle @joshcooper https://github.com/puppetlabs/puppet/pull/5799 has been opened separately to contain changes to encoding utility for isolated review. This PR now depends on that PR, both of which are ready for review.

puppetcla commented:
CLA signed by all contributors.

MosesMendoza commented:
No longer blocked.

This (in combination with #5799) has passed ad-hoc acceptance run, here: https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/ad-hoc/view/all/job/platform_puppet-agent_intn-van-sys_suite-ad-hoc-vmpooler-puppet-ad-hoc/15/

against these targets:
* redhat7-64a
* ubuntu1604-64a
* windows10ent-32a
* windows2008r2-64a
* windows2012r2-64a
* windows2012r2-6432a
* windows2012r2_ja-64a
* windows2012r2_ja-6432a
* windows2016-64a

MosesMendoza commented:
this will need rebased when #5799 is merged. That one has been retargeted at stable, but perhaps this should still go into master?

Iristyle commented on lib/puppet/util/log.rb:
```diff
@@ -170,6 +170,17 @@ def Log.with_destination(destination, &block)
+    else
+      message = _("Received a log message with invalid encoding:")
+      message << Puppet::Util::CharacterEncoding.convert_to_utf_8(msg.message.dump) << "\n"
+      # We only select the last 5 callers in the stack to avoid being spammy
```
5 was pretty arbitrary... I wonder if we should go a little deeper than that?

MosesMendoza replied:
Sure. Maybe 10 lines? I really wouldn't want to go deeper than 10.. Also just realized definitely need to make sure we don't access out of bounds of the array. Classic.

Iristyle replied:
I was actually looking at the bounds problem initially... it's not actually a problem, but you could also use `.first` for the same thing:

```
[1] pry(main)> foo = ['hi']
=> ["hi"]
[2] pry(main)> foo[0..4]
=> ["hi"]
[3] pry(main)> foo[0..4].join("\n")
=> "hi"
```

MosesMendoza replied:
i'm not sure whether to be grateful or dismayed

Iristyle commented on spec/unit/util/log_spec.rb:
```diff
@@ -18,6 +18,32 @@ def log_notice(message)
 
+  context "given a message with invalid encoding" do
+    let(:logs) { [] }
+    let(:invalid_message) { "\xFDfoo".force_encoding(Encoding::UTF_8) }
```
Maybe better to test with an invalid byte of another multi-byte encoding, like Shift-JIS?

```
[4] pry(main)> "\xFD\xFB".force_encoding(Encoding::Shift_JIS).valid_encoding?
=> false
```

MosesMendoza replied:
sure

MosesMendoza replied:
updated

Iristyle commented on lib/puppet/util/log.rb:
```diff
@@ -170,6 +170,17 @@ def Log.with_destination(destination, &block)
   def Log.newmessage(msg)
     return if @levels.index(msg.level) < @loglevel
 
+    if msg.message.valid_encoding?
```
In this case should we also bump up the `level`?

Our levels are:
* :debug
* :notice
* :info
* :warning
* :err

MosesMendoza replied:
I was on the fence about that... Since we're still logging the original message instead of a separate one re: the encoding I was hesitant to change too much about it - mostly to avoid surprising people and/or changing expectations/log levels in a (possibly?) backwards-incompatible way. I figured just logging at initial level was least surprising.

Iristyle replied:
I can buy that... for 💵

MosesMendoza commented on lib/puppet/util/log/destinations.rb:
```diff
@@ -212,37 +212,35 @@ def handle(msg)
 
+    # Attempt to convert our message to UTF-8
+    # Just returns a copy of the original message on failure to convert
+    message = Puppet::Util::CharacterEncoding.convert_to_utf_8(msg.to_s)
```
We don't need this anymore since we're doing it higher up

Iristyle replied:
👍

Iristyle commented on lib/puppet/util/log/destinations.rb:
```diff
@@ -212,6 +212,7 @@ def handle(msg)
 end
 
 Puppet::Util::Log.newdesttype :eventlog do
+  require 'puppet/util/character_encoding'
```
This `require` is no longer needed

MosesMendoza replied:
truth

Iristyle commented:
The hunch here is that this also fixes [PUP-7063](https://tickets.puppetlabs.com/browse/PUP-7063)

Iristyle commented:
A follow-on ticket has been filed at https://tickets.puppetlabs.com/browse/PUP-7880 to address encoding incompatibilities between `source` and `message` attributes of the log object.
